### PR TITLE
libavfi/dnn: add LibTorch as one of DNN backend

### PR DESCRIPTION
**Pull request by Semmer2**
PyTorch is an open source machine learning framework that accelerates
the path from research prototyping to production deployment. Official
websit: https://pytorch.org/. We call the C++ library of PyTorch as
LibTorch, the same below.

To build FFmpeg with LibTorch, please take following steps as reference:
1. download LibTorch C++ library in https://pytorch.org/get-started/locally/,
please select C++/Java for language, other options as your need.
2. unzip the file to your own dir, with command
unzip libtorch-shared-with-deps-latest.zip -d your_dir
3. export your_dir/libtorch/include and
your_dir/libtorch/include/torch/csrc/api/include to $PATH
export your_dir/libtorch/lib/ to $LD_LIBRARY_PATH
4. config FFmpeg with ../configure --enable-libtorch --extra-cflag=-I/your_dir/libtorch/include --extra-cflag=-I/your_dir/libtorch/include/torch/csrc/api/include --extra-ldflags=-L/your_dir/libtorch/lib/
5. make

To run FFmpeg DNN inference with LibTorch backend:
./ffmpeg -i input.jpg -vf dnn_processing=dnn_backend=torch:model=LibTorch_model.pt -y output.jpg
The LibTorch_model.pt can be generated by Python with torch.jit.trace() api.

### CODE REVIEW

File: libavfilter/dnn/dnn_backend_torch.cpp [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+#include "dnn_backend_common.h"
+#include "libavutil/opt.h"
+#include "queue.h"
+#include "safe_queue.h"
+}
+
+#define HEIGHT 360
+#define WIDTH 639
+#define UP_SCALE_TIMES 4
+
```
Comment (Semmer2) [inline libavfilter/dnn/dnn_backend_torch.cpp:42]: Fixed input size, still try to get input tensor for now.

Comment (Semmer2) [inline libavfilter/dnn/dnn_backend_torch.cpp:42]: Fixed input size and upscale times, still trying to get input tensor for now.

File: libavfilter/dnn/dnn_backend_torch.cpp [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+}
+
+static void th_free_request(THInferRequest *request)
+{
+    if (!request)
+        return;
+    if (request->input_tensor)
+        delete(request->input_tensor);
+    if (request->output)
+        delete(request->output);
```
Comment (Semmer2) [inline libavfilter/dnn/dnn_backend_torch.cpp:103]: Segmentation fault may occur when delete output tensor.

File: configure [hunk 3]
```diff
@@ -5304,9 +5306,9 @@ if test "$?" != 0; then
     die "C compiler test failed."
 fi
 
-add_cppflags -D_ISOC99_SOURCE
+add_cppflags -D_GLIBCXX_USE_CXX11_ABI=0 -D_ISOC99_SOURCE
```
Comment (xuguangxin) [inline configure:5336]: This will not accept by upstream. Why do we need to change this?

File: configure [hunk 4]
```diff
@@ -6489,6 +6491,7 @@ enabled libopus           && {
     }
 }
 enabled libpulse          && require_pkg_config libpulse libpulse pulse/pulseaudio.h pa_context_new
+enabled libtorch          && require_cpp libtorch torch/torch.h "torch::Tensor" -ltorch -lc10 -ltorch_cpu -lstdc++ -lpthread
```
Comment (xuguangxin) [inline configure:6538]: libtorch has package config file. better use it. https://pytorch.org/cppdocs/installing.html

File: libavfilter/dnn/dnn_backend_torch.cpp [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+        }
+        ff_safe_queue_destroy(th_model->request_queue);
+
+        while (ff_queue_size(th_model->lltask_queue) != 0) {
+            LastLevelTaskItem *item = (LastLevelTaskItem *)ff_queue_pop_front(th_model->lltask_queue);
+            av_freep(&item);
+        }
+        ff_queue_destroy(th_model->lltask_queue);
+
+        while (ff_queue_size(th_model->task_queue) != 0) {
```
Comment (xuguangxin) [inline libavfilter/dnn/dnn_backend_torch.cpp:490]: need ff_queue_destroy?

File: libavfilter/dnn_filter_common.c [hunk 6]
```diff
@@ -53,19 +53,31 @@ static char **separate_output_names(const char *expr, const char *val_sep, int *
...
-    if (!ctx->model_inputname) {
-        av_log(filter_ctx, AV_LOG_ERROR, "input name of the model network is not specified\n");
-        return AVERROR(EINVAL);
-    }
 
-    ctx->model_outputnames = separate_output_names(ctx->model_outputnames_string, "&", &ctx->nb_outputs);
-    if (!ctx->model_outputnames) {
-        av_log(filter_ctx, AV_LOG_ERROR, "could not parse model output names\n");
-        return AVERROR(EINVAL);
+    if (backend == DNN_TH) {
```
Comment (xuguangxin) [inline libavfilter/dnn_filter_common.c:63]: why tf and openvino not need this?

File: libavfilter/dnn_filter_common.c [hunk 7]
```diff
@@ -113,8 +125,11 @@ DNNReturnType ff_dnn_get_input(DnnContext *ctx, DNNData *input)
 
 DNNReturnType ff_dnn_get_output(DnnContext *ctx, int input_width, int input_height, int *output_width, int *output_height)
 {
-    return ctx->model->get_output(ctx->model->model, ctx->model_inputname, input_width, input_height,
-                                    (const char *)ctx->model_outputnames[0], output_width, output_height);
+    DNNBackendType backend = ctx->backend_type;
```
Comment (xuguangxin) [inline libavfilter/dnn_filter_common.c:128]: why pytroch is so speical ?

File: libavfilter/dnn/dnn_backend_torch.cpp [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+    int stride = frame->linesize[0];
+    uint8_t *src_line;
+    float *dst = NULL;
+    dst = (float*)malloc(size * 3 * sizeof(float));
+    if (!dst) {
+        delete(tensor);
+        tensor = NULL;
+        return;
+    }
+    //convert between packed and planner RGB24
```
Comment (xuguangxin) [inline libavfilter/dnn/dnn_backend_torch.cpp]: seems we have simliar function in swscale, gbr24ptopacked24. 
could we use swscale do this?

Comment (Semmer2) [reply libavfilter/dnn/dnn_backend_torch.cpp:231]: Will add code to support function like gbr24ptopacked24

Comment (Semmer2) [reply libavfilter/dnn/dnn_backend_torch.cpp]: > seems we have simliar function in swscale, gbr24ptopacked24. could we use swscale do this?

Will add code to support function like gbr24ptopacked24

File: libavfilter/dnn_filter_common.c (hunk 7 above)
Comment (Semmer2) [reply libavfilter/dnn_filter_common.c:128]: > why pytroch is so speical ?

Since pytorch C++ api do not need input/output names to do the inference.

**Comment by sramakintel** [comment]
Is this shared library or static? Seems to be shared as the package contains libtorch.so, any ways to build libtorch.a and integrate with FFmpeg?

**Pull request closed without merging**